src/algo/depth-to-rgb-calibration/optimizer.cpp

```diff
@@ -0,0 +1,1720 @@
+
+    if (data)
+    {
+        data->iteration_data_p.d_vals_x = interp_IDT_x;
+        data->iteration_data_p.d_vals_y = interp_IDT_y;
+        data->iteration_data_p.xy = rc.first;
+        data->iteration_data_p.rc = rc.second;
+    }
+        
+    res = calc_p_gradients( z_data, new_vertices, yuy_data, interp_IDT_x, interp_IDT_y, cal, p_mat, rc.second, rc.first, data );
+    return res;
+}
+
+std::pair<double, p_matrix> calc_cost_and_grad(
+    const z_frame_data & z_data,
+    const std::vector<double3>& new_vertices,
+    const yuy2_frame_data & yuy_data,
+    const calib & cal,
+    const p_matrix & p_mat,
+    data_collect * data = nullptr
+)
+{
+    auto uvmap = get_texture_map(new_vertices, cal, p_mat);
+    if( data )
+        data->iteration_data_p.uvmap = uvmap;
+
+    auto cost = calc_cost(z_data, yuy_data, uvmap, data ? &data->iteration_data_p.d_vals : nullptr );
+    auto grad = calc_gradients(z_data, new_vertices, yuy_data, uvmap, cal, p_mat, data);
+    return { cost, grad };
+}
+
+params::params()
+{
+    normalize_mat.vals[0] = 0.353692440000000;
+    normalize_mat.vals[1] = 0.266197740000000;
+    normalize_mat.vals[2] = 1.00926010000000;
+    normalize_mat.vals[3] = 0.000673204490000000;
+    normalize_mat.vals[4] = 0.355085250000000;
+    normalize_mat.vals[5] = 0.266275050000000;
+    normalize_mat.vals[6] = 1.01145800000000;
+    normalize_mat.vals[7] = 0.000675013750000000;
+    normalize_mat.vals[8] = 414.205570000000;
+    normalize_mat.vals[9] = 313.341060000000;
+    normalize_mat.vals[10] = 1187.34590000000;
+    normalize_mat.vals[11] = 0.791570250000000;
+
+    // NOTE: until we know the resolution, the current state is just the default!
+    // We need to get the depth and rgb resolutions to make final decisions!
+}
+svm_model_linear::svm_model_linear()
+{
+}
+svm_model_gaussian::svm_model_gaussian()
+{
+}
+void params::set_depth_resolution( size_t width, size_t height )
+{
+    AC_LOG( DEBUG, "... depth resolution= " << width << "x" << height );
+    // Some parameters are resolution-dependent
+    bool const XGA = (width == 1024 && height == 768);
+    if( XGA )
+    {
+        AC_LOG( DEBUG, "... changing IR threshold: " << grad_ir_threshold << " -> " << 2.5 << "  (because of resolution)" );
+        grad_ir_threshold = 2.5;
+    }
+}
+
+void params::set_rgb_resolution( size_t width, size_t height )
+{
+    AC_LOG( DEBUG, "... RGB resolution= " << width << "x" << height );
+}
+
+calib const & optimizer::get_calibration() const
+{
+    return _final_calibration;
+}
+
+rs2_dsm_params const & optimizer::get_dsm_params() const
+{
+    return _final_dsm_params;
+}
+
+double optimizer::get_cost() const
+{
+    return _params_curr.cost;
+}
+
+static
+void write_to_file( void const * data, size_t cb,
+    std::string const & dir,
+    char const * filename
+)
+{
+    std::string path = dir + '\\' + filename;
+    std::fstream f( path, std::ios::out | std::ios::binary );
```

Comment:
The behavior is not clear - should it append of override?
If it is append - then is there a method to clear its content?

Reply:
It overwrites if append isn't specified. ofstream default mode is just 'out'. I agree it's a little ambiguous, but not high priority right now -- it works in validation.